odl/tomo/geometry/detector.py

```diff
@@ -894,6 +895,640 @@ def __str__(self):
         return repr(self)
 
 
+class CylindricalDetector(Detector):
+
+    """A 2D detector on a cylindrical surface in 3D space.
+
+    The cylindrical surface that corresponds to the partition
+    is rotated to be aligned with given axes and
+    shifted to cross the origin. Note, the partition angle increases
+    in the clockwise direction, by analogy to flat detectors."""
+
+    def __init__(self, partition, axes, radius, check_bounds=True):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        partition : 2-dim. `RectPartition`
+            Partition of the parameter interval, corresponding to the
+            angular partition and height partition.
+        axes : sequence of `array-like`'s
+            Fixed pair of of unit vectors with which the detector is aligned.
+            The vectors must have shape ``(3,)`` and be linearly
+            independent.
+        radius : nonnegative float
+            Radius of the cylinder.
+        check_bounds : bool, optional
+            If ``True``, methods computing vectors check input arguments.
+            Checks are vectorized and add only a small overhead.
+
+        Examples
+        --------
+        Initialize a detector with height 8 and circle radius 2 extending to
+        90 degrees on both sides of the origin (a half cylinder).
+
+        >>> part = odl.uniform_partition(
+        ...     [-np.pi / 2, -4], [np.pi / 2, 4], [10, 8])
+        >>> det = CylindricalDetector(
+        ...     part, axes=[(1, 0, 0), (0, 0, 1)], radius = 2)
+        >>> det.axes
+        array([[ 1.,  0.,  0.],
+               [ 0.,  0.,  1.]])
+        >>> det.radius
+        2.0
+        >>> np.allclose(det.surface_normal([0, 0]), [ 0, -1,  0])
+        True
+        """
+        super(CylindricalDetector, self).__init__(partition, 3, check_bounds)
+        if self.ndim != 2:
+            raise ValueError('`partition` must be 2-dimensional, got ndim={}'
+                             ''.format(self.ndim))
+
+        axes, axes_in = np.asarray(axes, dtype=float), axes
+        if axes.shape != (2, 3):
+            raise ValueError('`axes` must be a sequence of 2 3-dimensional '
+                             'vectors, got {}'.format(axes_in))
+        if np.linalg.norm(np.cross(*axes)) == 0:
+            raise ValueError('`axes` {} are linearly dependent'
+                             ''.format(axes_in))
+
+        self.__axes = axes / np.linalg.norm(axes, axis=1, keepdims=True)
+
+        self.__radius = float(radius)
+        if self.__radius <= 0:
+            raise ValueError('`radius` must be positive')
+
+        initial_axes = np.array([[0, -1, 0], [0, 0, 1]])
+        r1 = rotation_matrix_from_to(initial_axes[0], axes[0])
+        r2 = rotation_matrix_from_to(initial_axes[1], axes[1])
+        self.__rotation_matrix = np.matmul(r1, r2)
```

Comment:
Are you sure this is correct?

First, if `axes` are not perpendicular, how can you ever rotate the `inital_axes` to `axes`? To avoid getting insanely complex geometries with sheared cylinders, I think we should ensure that the axes are perpendicular.

And for perpendicular `axes`, I would expect that one first has to rotate one of the two vectors to its final location, and then *around* that new axis to match the other one.

Please make sure with some with some checks (e.g. unit tests) that this part is correct.

Reply:
You are right, I don't know what I was thinking here...

Reply:
So, it should be like this
r1 = rotation_matrix_from_to(initial_axes[0], axes[0])
r2 = rotation_matrix_from_to(np.matmul(r1, initial_axes[1]), axes[1])
and it will work, if axes are perpendicular.